packages/view-types/statistical-plots/src/DotPlot.js

```diff
@@ -0,0 +1,131 @@
+import React from 'react';
+import clamp from 'lodash/clamp';
+import { VegaPlot, VEGA_THEMES } from '@vitessce/vega';
+import { capitalize } from '@vitessce/utils';
+import plur from 'plur';
+
+/**
+ * Gene expression dot plot,
+ * implemented with the VegaPlot component.
+ * @param {object} props
+ * @param {object[]} props.data The expression data, an array
+ * of objects with properties `value`, `gene`, and `set`.
+ * @param {string} props.theme The name of the current Vitessce theme.
+ * @param {number} props.width The container width.
+ * @param {number} props.height The container height.
+ * @param {number} props.marginRight The size of the margin
+ * on the right side of the plot, to account for the vega menu button.
+ * By default, 90.
+ * @param {number} props.marginBottom The size of the margin
+ * on the bottom of the plot, to account for long x-axis labels.
+ * Default is allowing the component to automatically determine the margin.
+ * @param {string|null} props.featureValueTransformName A name
+ * for the feature value transformation function.
+ */
+export default function DotPlot(props) {
+  const {
+    data: rawData,
+    theme,
+    width,
+    height,
+    marginRight,
+    marginBottom,
+    obsType,
+    keyLength = 36,
+    featureType,
+    featureValueType,
+    featureValueTransformName,
+  } = props;
+
+  // Add a property `keyGroup` and `keyFeature` which concatenates the key and the name,
+  // which is both unique and can easily be converted
+  // back to the name by taking a substring.
+  const data = rawData.map(d => ({
+    ...d,
+    keyGroup: d.groupKey + d.group,
+    keyFeature: d.featureKey + d.feature,
+  }));
+
+  // Get the max characters in an axis label for autsizing the bottom margin.
+  const maxCharactersForGroup = data.reduce((acc, val) => {
+    // eslint-disable-next-line no-param-reassign
+    acc = acc === undefined || val.group.length > acc ? val.group.length : acc;
+    return acc;
+  }, 0);
+  const maxCharactersForFeature = data.reduce((acc, val) => {
+    // eslint-disable-next-line no-param-reassign
+    acc = acc === undefined || val.feature.length > acc ? val.feature.length : acc;
+    return acc;
+  }, 0);
+  // Use a square-root term because the angle of the labels is 45 degrees (see below)
+  // so the perpendicular distance to the bottom of the labels is proportional to the
+  // square root of the length of the labels along the imaginary hypotenuse.
+  // 30 is an estimate of the pixel size of a given character and seems to work well.
+  const autoMarginVertical = marginBottom
+    || 30 + Math.sqrt(maxCharactersForFeature / 2) * 30;
+  const autoMarginHorizontal = marginRight
+    || 30 + Math.sqrt(maxCharactersForGroup / 2) * 30;
+
+  const plotWidth = clamp(width - autoMarginHorizontal - 120, 10, Infinity);
+  const plotHeight = clamp(height - autoMarginVertical, 10, Infinity);
```

Comment:
This trick is used elsewhere, no? Should it refactored into its own hook or function?